### ISSUE TITLE
Support for 1D and Non-square convolutional kernels

### ISSUE THREAD

**Issue by brazilla** [hughperkins/DeepCL#37]
Currently only 2D square convolutional kernels are supported.
Thx.

**Comment by hughperkins** [comment]
Understood.  Involves a fair amount of work.  Will ponder.  Anyone else who arrives in this thread, and wants non-square kernels, please add a `+1` into this thread, to motivate me a bit :-)

**Comment by hughperkins** [comment]
Oh, hmmm, kernels.... is this for Inception?

**Comment by brazilla** [comment]
Mainly for NLP and other sequence processing.

**Comment by hughperkins** [comment]
Ah, hmmm, right.

Presumably you'd want the input 'images' to also be non-squared too?  Presumably, quite very not-square, like 1-dimensional?

**Comment by brazilla** [comment]
Yeah,exactly.

**Comment by hughperkins** [comment]
Hmmm... does sound potentially quite useful generally.  By the way, just out of curiosity do you know any papers that use this technique?

**Comment by brazilla** [comment]
http://nal.co/papers/Kalchbrenner_DCNN_ACL14
nlp.stanford.edu/~socherr/EMNLP2013_RNTN.pdf
http://www-nlp.stanford.edu/pubs/SocherLinNgManning_ICML2011.pdf

**Comment by hughperkins** [comment]
Thanks!

**Comment by rhyzomatic** [comment]
+1. Has there been any progress with this?

**Comment by hughperkins** [comment]
Hmmmm .... well.... clnn contains a wrapper for 1-d convolutions.  https://github.com/hughperkins/clnn/blob/master/TemporalConvolution2.lua  I confess I'm unlikely to dedicate any time to implementing this in DeepCL in the short-term.  I dont imagine it's tons of work, as long as you:
- make it *only* for 1-d, ie not generally rectangular support
- use the existing im2col convolution implementation.  It's fast, and means you wont need to roll up your sleeves and dive into opencl too much (might need to ... a bit... for im2col ... maybe :-P)
It is however non-trivial enough that I'm not just going to spend a few hours one morning, and make it work.  It's probably about a weeks work, I would think, ballpark.  40 man-hours or so.

**Comment by hughperkins** [comment]
Note: I think an appropriate way forward for this might be to migrate to use https://github.com/BVLC/caffe/issues/4155 , so that any kernels are shared across OpenCL projects.

**Comment by dex1990** [comment]
hi,@hughperkins
did DeepCL support 3D square convolutional layers ? e.g. 3x3x64, which means size of filter is width 3,height 3 and depth 64.

**Comment by hughperkins** [comment]
Correct. Can you confirm what makes you feel this might not be the case?

On 15 November 2016 10:35:01 GMT+00:00, dex1990 <notifications@github.com> wrote:
>hi,@hughperkins
>did DeepCL support 3D square convolutional layers ? e.g. 3x3x64, which
>means size of filter is width 3,height 3 and depth 64.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/hughperkins/DeepCL/issues/37#issuecomment-260606541

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by dex1990** [comment]
Thank you for your answer.
I have not found any example on it, maybe i missed some relevant examples.
I have trained model from this project:
   https://github.com/huangzehao/caffe-vdsr
some layers of this network have input size: 40x40x64 and 64 filters with size: 3x3x64, so the output size is 40x40x64. I want to test the trained network with DeepCL.

**Comment by 0StackOverflow0** [comment]
I probably misunderstand, but wouldn't the 2D implementation work for 1D as well?

Would you not stack the input? (i.e. LxWxD 1x1xN). You'd get a scalar from each filter, which I believe is the expectation.

I'm interested in adding 3D convolutional filters (once I do some more reading), but I'm hoping to have a proposal for stride soon.

**Comment by hughperkins** [comment]
> I'm hoping to have a proposal for stride in the not too distant future.

Sounds good :)

> I wonder how hard 1D would be after doing that.

Its not conceptually hard, but it might involve a certain amount of legwork :)
